File: code/ex1003.java
import java.util.Stack;

class Solution {
    public boolean isValid(String S) {
        Stack<Character> stack = new Stack<>();
        
        for (int i = 0; i < S.length(); i++) {
            stack.push(S.charAt(i));
            if (S.charAt(i) == 'c') {
                if (stack.size() < 3 || stack.get(stack.size() - 2) != 'b' || stack.get(stack.size() - 3) != 'a')    return false;
                else {stack.pop(); stack.pop(); stack.pop();}
            }
        }
        
        return (stack.size() == 0) ? true : false;
    }
}

File: code/ex1003.cpp
// Stack

class Solution {
public:
    bool isValid(string S) {
        int n = S.length();
        
        stack<char> stk;
        for (int i = 0; i < n; i++) {
            if (S[i] == 'c') {
                if (stk.size() < 2)    return false;
                char p1 = stk.top();    stk.pop();
                char p2 = stk.top();    stk.pop();
                if (!(p1 == 'b' && p2 == 'a'))    return false;
            } else {
                stk.push(S[i]);
            }
        }
        
        return stk.empty();
    }
};